src/workspace/discovery.rs

```diff
@@ -391,29 +399,94 @@ fn discover_workspace_config_files_for_single_dir(
+                }
+                .into(),
+              );
+            }
+            validate_member_url_is_descendant(&member_dir_url)?;
             let member_config_folder =
-              find_member_config_folder(raw_member, &member_dir_url)?;
-            add_member(raw_member, member_dir_url, member_config_folder)?;
+              find_member_config_folder(&member_dir_url)?;
+            let previous_member = final_members
+              .insert(new_rc(member_dir_url), member_config_folder);
+            if previous_member.is_some() {
+              return Err(
+                ResolveWorkspaceMemberError::Duplicate {
+                  member: raw_member.to_string(),
+                }
+                .into(),
+              );
+            }
           }
         }
       }
       if let Some(pkg_json) = root_config_folder.pkg_json() {
         if let Some(members) = &pkg_json.workspaces {
-          let mut has_warned = false;
-          for raw_member in members {
-            if raw_member.contains('*') {
-              if !has_warned {
-                has_warned = true;
-                log::warn!(
-                  "Wildcards in npm workspaces are not yet supported. Ignoring."
-                );
-              }
-              continue;
-            }
+          let (pattern_members, path_members): (Vec<_>, Vec<_>) =
+            members.iter().partition(|member| {
+              is_glob_pattern(member) || member.starts_with('!')
+            });
+          let patterns = pattern_members
+            .iter()
+            .map(|raw_member| {
+              PathOrPattern::from_relative(
+                pkg_json.dir_path(),
+                &format!("{}package.json", ensure_trailing_slash(raw_member)),
+              )
+              .map_err(|err| {
+                ResolveWorkspaceMemberError::NpmMemberToPattern {
+                  base: root_config_file_directory_url.clone(),
+                  member: raw_member.to_string(),
+                  source: err,
+                }
+              })
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+          let pkg_json_paths = if patterns.is_empty() {
+            Vec::new()
+          } else {
+            FileCollector::new(|_| true)
+              .ignore_git_folder()
+              .ignore_node_modules()
+              .set_vendor_folder(maybe_vendor_dir.clone())
+              .collect_file_patterns(
+                opts.fs,
+                FilePatterns {
+                  base: pkg_json.dir_path().to_path_buf(),
+                  include: Some(PathOrPatternSet::new(patterns)),
+                  exclude: PathOrPatternSet::new(Vec::new()),
+                },
+              )
+              .map_err(|err| {
+                WorkspaceDiscoverErrorKind::FailedCollectingNpmMembers {
+                  package_json_url: pkg_json.specifier(),
+                  source: err,
+                }
+              })?
```

Comment:
Not sure if this is exactly correct, but should be good enough for a first pass.